### ISSUE TITLE
domain.c assertion failure

### ISSUE THREAD

**Issue by gasche** [ocaml/ocaml#11800]
After merging #11187 I received an email from Github about a test failure on the linux-debug CI configuration, which runs the testsuite with the debug runtime.

[link to failed job](https://github.com/ocaml/ocaml/actions/runs/3641094040/jobs/6146721242)

relevant log:
```
... testing 'weaktest_par_load.ml' with 1 (native) Process 48603 got signal 6(Aborted), core dumped
[...]
Commandline: /home/runner/work/ocaml/ocaml/testsuite/tests/weak-ephe-final/_ocamltest/tests/weak-ephe-final/weaktest_par_load/ocamlopt.byte/weaktest_par_load.opt
### begin stdout ###
[01] file runtime/domain.c; line 992 ### Assertion failed: di->backup_thread_msg == BT_INIT || di->backup_thread_msg == BT_TERMINATE
### end stdout ###
```

This looks like a real issue (probably a bug) that only shows up very rarely as a CI failure in debug mode.

The code where the assert fails is the code setting up the backup thread:

https://github.com/ocaml/ocaml/blob/97aa649e3c26e5f9ba2e3e5fc88115bc60afa5f9/runtime/domain.c#L984-L993

The test where the failure occurred:

https://github.com/ocaml/ocaml/blob/97aa649e3c26e5f9ba2e3e5fc88115bc60afa5f9/testsuite/tests/weak-ephe-final/weaktest_par_load.ml

(cc @sadiqj @Engil @kayceesrk)

**Comment by gasche** [comment]
I tried to think about how failing this assertion is possible. Here is a proposed scenario in the middle of a running program:

1. the main domain spawns a domain A
2. A starts its backup thread
3. A terminates, and in the process:
   - it sets `domain_self->backup_thread_running = 0`
   - it has not signalled the backup thread to terminate yet
   - it releases `all_domains_lock`, and is about to enter the delicate part that collects the rest of the state without owning `domain_self` anymore
4. the main domain spawns a domain B, which reuses the slot of A
5. B starts its backup thread

At this point, if my reasoning is right, `B` sees `domain_self->backup_thread_running == 0`, but the backup thread of A has not terminated yet, and the assumption may fail.

## More details

`domain_terminate` makes `domain_self` available for reuse:

https://github.com/ocaml/ocaml/blob/97aa649e3c26e5f9ba2e3e5fc88115bc60afa5f9/runtime/domain.c#L1698-L1722

(notice that this sets `backup_thread_running` back to 0)

`domain_terminate` signals the backup thread only much later:

https://github.com/ocaml/ocaml/blob/97aa649e3c26e5f9ba2e3e5fc88115bc60afa5f9/runtime/domain.c#L1753-L1758

meanwhile, a concurrent `domain_create` call may have ran `install_backup_thread` and failed on the assertion:

https://github.com/ocaml/ocaml/blob/97aa649e3c26e5f9ba2e3e5fc88115bc60afa5f9/runtime/domain.c#L984-L993

**Comment by gasche** [comment]
Note: it looks like this kind of failure scenarios should have been prevented by commit https://github.com/ocaml/ocaml/commit/a8e8ff30b91fe87b778adc71f2e59d3cfb696f6d from @kayceesrk in June 2020, which moves the backup-thread-termination logic within the core loop of `domain_terminate`. But I think that it was broken again by https://github.com/ocaml/ocaml/commit/faeaeb1384de7173b6b95648a3ebd4303b5356a3 in July 2020.

The June 2020 commit put in the core terminate loop a signal to the backup thread, followed by busy-waiting until the backup thread at terminate. The assertion in `install_backup_thread` was thus very strong, the status must be `BT_INIT` because we are sure that the backup thread terminated.

I believe that the intention of the July 2020 commit was to remove the busy-waiting loop, and instead relax the `install_backup_thread` invariant to see either `BT_INIT` or `BT_TERMINATE` (and busy-wait there in the second case). Unfortunately, the BT_TERMINATE write to the status was moved back to the end of domain_terminatee, after domain reuse becomes possible, re-introducing the bug.

If this analysis is correct, the fix should be to signal termination to the backup thread within the core termination loop, but not busy-wait.

**Comment by abbysmal** [comment]
I think your analysis is quite on point, and I don't see anything wrong with moving termination signaling from within the part were we still hold the domain lock. (and `all_domain_lock`)

Do you want me to submit a fix or are you working on one?

Good job on figuring this one out. :-)

**Comment by gasche** [comment]
I was indeed working on a PR, #11801 :-)

**Comment by gasche** [comment]
I closed #11801 as i realized that I don't understand the code enough to be confident in a fix. My summary of the situation there is also relevant here;

> Let me try to explain what is bugging me about `domain_lock`, maybe the experts in the room can let me understand better. If I read the code correctly, the part of domain_terminate that writes the BT_TERMINATE status is done while holding the domain lock, and releases the domain lock right away:
> 
> https://github.com/ocaml/ocaml/blob/97aa649e3c26e5f9ba2e3e5fc88115bc60afa5f9/runtime/domain.c#L1753-L1758
> 
> On the other hand, `create_domain` takes the domain lock:
> 
> https://github.com/ocaml/ocaml/blob/97aa649e3c26e5f9ba2e3e5fc88115bc60afa5f9/runtime/domain.c#L583-L588
> 
> and the call to `install_backup_thread` only occurs after this `domain_create` call has taken the lock:
> 
> https://github.com/ocaml/ocaml/blob/97aa649e3c26e5f9ba2e3e5fc88115bc60afa5f9/runtime/domain.c#L1081-L1082
> 
> Naively this suggests that the race scenario I described above, when creating this PR, cannot actually happen: a new domain reusing the same `dom_internal` slot could only install its backup thread after the previous domain termination has released the domain lock, and this occurs after `BT_TERMINATE` has been written, so at this point the `install_backup_thread` assertion failure is not possible anymore.
> 
> But: there are other uses of the domain lock, and it may be that something funny happens. (Could something in `domain_terminate` release and re-acquire the domain lock after leaving the STW set but before writing the BT_TERMINATE status?)
> 
> I must say that it's not so easy to follow the use of the various locking mechanisms in domain.c and the invariants they should protect. For example the backup thread sometimes waits on an interruptor-related condition variable `s->cond`, sometimes on `domain_cond`, and it is not obvious why or what assumptions are made. Independently, the part of `domain_terminate` after removal from the STW set seems careful not to access `domain_self`, only `domain_state`... except precisely for the three lines that deal with the backup thread.
> 
> Side question: online documentation on condition variable suggests to use a `while (cond-is-false) pthread_cond_wait(cond)` approach, to avoid spurious wakeups. I wonder why the backup thread code uses a simple `if (cond is false) pthread_cond_wait(cond)` instead, what happens in the case of spurious wakeups?

**Comment by kayceesrk** [comment]
It is very surprising that this assertion failure occurs. I'm trying to recreate this locally on the failing list. But I've not had success so far.

**Comment by gasche** [comment]
I believe that the mystery was solved by @jmid in #12707, so this can now be closed.
